[1.21.6][B3D][Breaking] Extend enums to cover most or all GL values

RogueLogix opened the pull request:
- TextureFormat extended to cover most formats. VkFormat naming convention used because they are more specific/consistent than the three part OpenGL format specification.
- SourceFactor/DestFactor completed to match GL's options
- Storage Texel Buffer (STB) and Shader Storage Buffer Object (SSBO) uniform types defined
- Remaining GL shader stages defined.
- Remaining sampler address modes added
- Remaining depth test functions added
- GpuBuffer usage bits for indirect  STB, SSBO, and transform feedback added, as well as 8 bits reserved for implementation specific uses
- GpuTexture usage bit for storage image added, as well as 8 bits reserved for implementation specific uses

Two vanilla bugs are also fixed
1) Usage of `GL_DEPTH_COMPONENT32` is replaced with `GL_DEPTH_COMPONENT32F` as only support for the latter is required by GL32C spec. This was remapped to `D32_SFLOAT`
2) `RED8I` used `GL_UNSIGNED_BYTE` host data type, which is inconsistent with the actual data type of `GL_BYTE`. This was remapped to `R8_SINT` which uses `GL_BYTE`

RogueLogix committed:
Extend TextureFormat to cover most texture formats.

RogueLogix committed:
Add MIRRORED_REPEAT and CLAMP_TO_BORDER AddressModes

RogueLogix committed:
Add other ShaderType stages

RogueLogix committed:
Add remaining depth test options

RogueLogix committed:
Add remaining logic op options

RogueLogix committed:
Add POINT polygon mode

RogueLogix committed:
Add remaining Source and Dest blend factors

RogueLogix committed:
remove USCALED and SSCALED formats

RogueLogix committed:
Add additional GpuBuffer usages, and reserve range for implementation defined usages in bitmask

RogueLogix committed:
Add additional GpuTexture usages, and reserve range for implementation defined usages in bitmask

RogueLogix committed:
Add STB and SSBO uniform types

RogueLogix committed:
Replace usage of not-required GL_DEPTH_COMPONENT32 with required by spec GL_DEPTH_COMPONENT32F

RogueLogix committed:
Remove USCALED/SSCALED 10/10/10/2 formats
Also remove minimally supported anywhere D16S8, 64bit, and signed 10/10/10/2 formats

RogueLogix committed:
Fix style errors

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=2359) <br/>All committers have signed the CLA.

RogueLogix commented on patches/com/mojang/blaze3d/opengl/GlProgram.java.patch:
```diff
@@ -0,0 +1,10 @@
+                         yield new Uniform.Utb(k, l, Objects.requireNonNull(renderpipeline$uniformdescription.textureFormat()));
+                     }
+                 }
++                default -> null;
```
Implementation of Stb/Ssbo uniform types will be done in a separate PR

RogueLogix committed:
Add FRAME_TRANSIENT and ONE_TIME_USE GpuBuffer usage bits

Technici4n commented:
Shouldn't we add selected useful values, rather than _all of them_?

I would also put a comment above the added enum entries to make it clear that they are NeoForge extensions.

XFactHD commented:
I agree with Tech on this. I would let `AddressMode`, `DepthTestFunction`, `UniformType` and maybe `ShaderType` slide but the rest of the enum extensions will need proper justification.
The usage bits seem fine at a glance (assuming they are actually implemented) but need some proper documentation on what they are for and what kind of limitations they have (i.e. extensions to check for).
Extensions which are either specific to a hardware vendor (i.e. task and mesh shaders) or aren't supported by OpenGL at all (i.e. one of the added texture formats) seem very problematic to me. Also, we cannot under any circumstances rename *any* vanilla enum entries which, unless I have overlooked something, was done for the texture formats.

RogueLogix commented:
@Technici4n 
> Shouldn't we add selected useful values, rather than _all of them_?

Not necessarily. For `AddressMode`, `DepthTestFunction`, `LogicOp`, `SourceFactor`/`DestFactor`, and `PolygonMode`, nothing further than adding the enum value (and respective switch in `GlConst` is required.
Adding one option to them is no more an work than adding all. Adding all options means less reason to go around B3D to get that value.

`TextureFormat` (except the compressed ones) also fall into this boat, but i can par it down at least some.
Some are mostly duplicates (ie: BGR vs RGB), and some others (most) hardware/drivers dont really like anyway.

On the compressed ones specific. Support for those formats is not part of the OpenGL core spec, but is ubiquitous (with even MacOS having support via OpenGL). In my opinion, use of them should be encouraged for large textures, as they use substantially less vram. However, maybe they shouldn't be supported through B3D at all.

> I would also put a comment above the added enum entries to make it clear that they are NeoForge extensions.

Will add.

@XFactHD 
> the rest of the enum extensions will need proper justification.

Of the remaining enums, as mentioned above, `LogicOp`, `SourceFactor`/`DestFactor`, and `PolygonMode` require no further work past defining the enum and adding an entry to their respective switch in `GlConst`. Adding the remaining options here removes a reason to go around B3D for state it already has some representation of.

`TextureFormat` is largely the same idea, and was actually the on I was running into a problem with, not having any format that would work for the data I needed to put into a UTB. Adding just `R32G32B32A32_UINT` or `R32G32B32A32_SINT` would solve that specific case, but also wouldn't in many other cases so I felt it better to avoid this same situation in the future.

`ShaderType` being a maybe, I'll explain anyway. I don't expect Neo to support any of the added stages _except_ compute. The others do have some amount of support from OpenGL (even if that is a currently vendor extension). I felt that adding the other options for alternate backends to implement was a better option than `IExtensibleEnum` or forcing an implementation specific enum for API generic stages. 

> the usage bits seem fine at a glance (assuming they are actually implemented) but need some proper documentation on what they are for and what kind of limitations they have (i.e. extensions to check for).

Will add.

>  task and mesh shaders

It's a vendor extension, but is also supported by AMD hardware. It's also not always a vendor extension (VK_EXT_mesh_shader), with [GL_EXT_mesh_shader in the works too](https://github.com/KhronosGroup/OpenGL-Registry/pull/640).
<https://opengl.gpuinfo.org/listreports.php?extension=GL_NV_mesh_shader> (if an AMD card isn't on the first page, searching `AMD` for renderer will find some).

> one of the added texture formats

Three, actually. `R8_SRGB`, `R8G8_SRGB`, and `R4G4_UNORM_PACK8` (which is probably the one you noticed, given its at the top).

I left `R8_SRGB`/`R8G8_SRGB` because they are the less component variants of `R8G8B8_SRGB` and `R8G8B8A8_SRGB` which is supported. Though SRGB encoding is unlikely to be used, maybe all of those should be removed entirely.

While OpenGL has no way to define `R4G4_UNORM_PACK8`, it is pretty widely supported by Vulkan. Though this specific format could also be unpacked in the shader (which is potentially what the driver is doing anyway) with `uvec2((packed >> 4) & 0xF, packed & 0xF)` and could be removed anyway because that is a reasonable alternative through `R8_UINT`

>  ... cannot under any circumstances rename any vanilla enum entries

Will flip the alias

RogueLogix committed:
Add comments to all extended enums/usages

RogueLogix committed:
Reduce TextureFormat count

Nearly duplicate formats reduced to one variant, most widely supported by hardware variant picked.
Commonly unavailable in hardware formats (3 component) removed.
SRGB formats removed.
Name alias flipped to keep vanilla constants in the enum declaration.

RogueLogix committed:
move LogicOp CLEAR and NONE to the correct locations within the enum

RogueLogix committed:
correct comments in GpuBuffer

Technici4n commented on patches/com/mojang/blaze3d/platform/LogicOp.java.patch:
```diff
@@ -0,0 +1,26 @@
+ public enum LogicOp {
++    CLEAR,
++    OR_REVERSE,
++    // Neo: Extended logic ops
```
Weird placement of the constants and the comment here.

RogueLogix replied:
looks like i forgot to push up fixing this last night

Technici4n commented on patches/com/mojang/blaze3d/opengl/GlCommandEncoder.java.patch:
```diff
@@ -88,6 +86,45 @@
++                    case CLEAR -> org.lwjgl.opengl.GL11C.GL_CLEAR;
++                    case AND -> org.lwjgl.opengl.GL11C.GL_AND;
++                    case AND_REVERSE -> org.lwjgl.opengl.GL11C.GL_AND_REVERSE;
++                    case NONE -> org.lwjgl.opengl.GL11C.GL_COPY;
```
NONE here?

RogueLogix replied:
Its not a required case, the if check does make it unreachable, but does also describe the behavior correctly.
 `GL_COPY` is a "take source" op, or as if logic-op is disabled.
<https://www.khronos.org/opengl/wiki/Logical_Operation>

Technici4n commented on patches/com/mojang/blaze3d/opengl/GlConst.java.patch:
```diff
@@ -1,31 +1,266 @@
+             case SRC_COLOR -> 768;
+             case ZERO -> 0;
++            case SRC_ALPHA_SATURATE -> org.lwjgl.opengl.GL11C.GL_SRC_ALPHA_SATURATE;
++            case SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_SRC1_COLOR;
++            case ONE_MINUS_SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_COLOR;
++            case SRC1_ALPHA -> org.lwjgl.opengl.GL15C.GL_SRC1_ALPHA;
++            case ONE_MINUS_SRC1_ALPHA -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_ALPHA;
```
What about the GL33 ones here?

RogueLogix replied:
This probably slipped through because GL3xC.

Requirement here is GL 3.3 or GL_ARB_blend_func_extended
Can add a comment about the requirement to the enum definition, and later will also need to add a capability check.

RogueLogix replied:
javadoc added to sourcefactor/destfactor on this

Technici4n commented on patches/com/mojang/blaze3d/opengl/GlConst.java.patch:
```diff
@@ -1,31 +1,266 @@
+             case SRC_ALPHA_SATURATE -> 776;
+             case SRC_COLOR -> 768;
+             case ZERO -> 0;
++            case SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_SRC1_COLOR;
++            case ONE_MINUS_SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_COLOR;
++            case SRC1_ALPHA -> org.lwjgl.opengl.GL15C.GL_SRC1_ALPHA;
++            case ONE_MINUS_SRC1_ALPHA -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_ALPHA;
```
GL33 as well.

XFactHD commented on patches/com/mojang/blaze3d/buffers/GpuBuffer.java.patch:
```diff
@@ -0,0 +1,26 @@
+     public static final int USAGE_UNIFORM = 128;
+     public static final int USAGE_UNIFORM_TEXEL_BUFFER = 256;
++    // Neo: Extended buffer usage, some may not always be supported, check backend capabilities before use
++    public static final int USAGE_TRANSFORM_FEEDBACK = USAGE_UNIFORM_TEXEL_BUFFER << 1;
++    // Neo: Requires OpenGL 4.2, GL_ARB_shader_image_load_store, or equivalent
++    public static final int USAGE_STORAGE_TEXEL_BUFFER = USAGE_TRANSFORM_FEEDBACK << 1;
++    // Neo: Requires OpenGL 4.3, GL_ARB_shader_storage_buffer_object, or equivalent
++    public static final int USAGE_STORAGE_BUFFER = USAGE_STORAGE_TEXEL_BUFFER << 1;
++    // Neo: Requires OpenGL 4.0, GL_ARB_draw_indirect, or equivalent
++    public static final int USAGE_INDIRECT = USAGE_STORAGE_BUFFER << 1;
++    // Neo: Requires that this buffer be uploaded to before being used for a given frame
++    //      use of this buffer without prior upload on a given frame may crash
++    public static final int USAGE_PERF_FRAME_TRANSIENT = USAGE_INDIRECT << 1;
++    // Neo: Hints that the contents of this buffer will be used once after upload
++    //      requires use of USAGE_PERF_FRAME_TRANSIENT
++    //      repeated use of the data is valid, this is a performance hint
++    public static final int USAGE_HINT_ONE_TIME_USE = USAGE_PERF_FRAME_TRANSIENT << 1 | USAGE_PERF_FRAME_TRANSIENT;
++    // Neo: Top 8 bits reserved for implementation specific usages
++    public static final int USAGE_FIRST_IMPLEMENTATION_DEFINED = 1 << 24;
```
These comments should be javadocs instead and the first field needs some details on what it's used for.

RogueLogix replied:
> first field needs some details on what it's used for.

The comments on the other fields (except bit reservation) are there specifically for the GL version/extension requirements, and not their use. Transform feedback is GL 3.0 and always supported, hence no comment. If use comments should be added, should they be added for all (including vanilla)?

XFactHD replied:
Use comments should definitely be added to fields that are patched in. The vanilla fields can be handled by Parchment.

XFactHD commented on patches/com/mojang/blaze3d/textures/GpuTexture.java.patch:
```diff
@@ -0,0 +1,14 @@
++    // Neo: Requires OpenGL 4.2, ARB_shader_image_load_store, or equivalent
++    public static final int USAGE_STORAGE = USAGE_CUBEMAP_COMPATIBLE << 1;
++    // Neo: Top 8 bits reserved for implementation specific usages
++    public static final int USAGE_FIRST_IMPLEMENTATION_DEFINED = 1 << 24;
```
These comments should be javadocs instead.

XFactHD commented on patches/com/mojang/blaze3d/platform/PolygonMode.java.patch:
```diff
@@ -0,0 +1,12 @@
+-    WIREFRAME;
++    WIREFRAME,
++    // Neo: Extended polygon mode
++    POINT,
```
What's the use case of this? Wireframe is already a rare encounter (my understanding is that it's almost exclusively used for debugging) and this seems like something the would be even less relevant/useful.

RogueLogix replied:
POINT rendering in general (not the polygon mode specifically, but the primitive topology as a whole) is generally useful for particle systems. That is probably a rare enough usecase to not warrant adding this here.

To support POINT rendering, it would probably actually be a better idea to add a POINT mode to VertexFormat.Mode instead.

I'll remove this.

RogueLogix committed:
Better comments on GpuBuffer usages

RogueLogix committed:
Remove R4G4_UNORM_PACK8, S8_UNORM, and compressed formats.

RogueLogix committed:
Add javadoc for TextureFormats. Remove all packed formats, and low precision depth formats.

RogueLogix committed:
Remove POINT polygon mode.

RogueLogix committed:
Add javadoc to GpuTexture usages.

RogueLogix committed:
Add javadoc to SourceFactor/DestFactor

RogueLogix committed:
Remove dead code and remove unused ClientHooks import

RogueLogix committed:
Add javadoc to ShaderType

RogueLogix committed:
Remove tess/geo shaders from GlConst

RogueLogix committed:
Remove unnecessary hunk from GlConst

XFactHD commented on patches/com/mojang/blaze3d/shaders/ShaderType.java.patch:
```diff
@@ -0,0 +1,28 @@
++     */
++    TASK("task", ".tsh"),
++    /**
++     * @see ShaderType#MESH
```
I assume this should link back to `TASK` and not `MESH`

RogueLogix replied:
_yes, yes it should_

shartte commented:
Just conceptually, why is extending the enums a breaking change?

RogueLogix commented:
It is a breaking change for any implementation of B3D, as previously unknown values could now be passed through the enums parameters and/or usage bits.
If any change to the built-in implementation is required (in this PR's case, the changes to `GlConst`), a similar change is required in all other implementations.
For the same reason, adding or modifying any function to any B3D interface or class without that being default handled within the API definition is similarly a breaking for B3D implementations.

This, however, is _not_ breaking for users of the API (except the removal of D24S8).

RogueLogix committed:
Fix mesh shader javadoc link
